docs/source/design/user_messaging.rst

```diff
@@ -0,0 +1,120 @@
+.. -*- mode: rst -*-
+.. vi: set ft=rst sts=4 ts=4 sw=4 et tw=79:
+
+.. _chap_design_user_messaging:
+
+*******************************************************
+User messaging: result records vs exceptions vs logging
+*******************************************************
+
+.. topic:: Specification scope and status
+
+   This specification provides a partial overview of the implementation goal.
+
+Motivation
+==========
+
+This specification delineates the applicable contexts for using
+:ref:`result records <chap_design_result_records>`, :ref:`exceptions <chap_design_exception_handling>`,
+:ref:`progress reporting <chap_design_progress_reporting>`, specific :ref:`log levels <chap_design_log_levels>`,
+or other types of user messaging processes.
+
+
+Specification
+=============
+
+Result records
+--------------
+
+**Result records are the only return value format** for all DataLad interfaces.
+
+Contrasting with classic Python interfaces that return specific non-annotated values,
+DataLad interfaces (i.e. subclasses of :py:class:`datalad.interface.base.Interface`)
+implement message passing by yielding :ref:`result records <chap_design_result_records>`
+that are associated with individual operations. Result records are routinely inspected throughout
+the code base and their annotations are used to inform general program flow and error handling.
+
+Command calls can include an ``on_failure`` parameterization to specify how to
+proceed with a particular operation if a returned result record is
+:ref:`classified as a failure result <target-result-status>`. Command calls can
+also include a ``result_renderer`` parameterization to explicitly enable or
+disable the rendering of result records.
+
+Developers should be aware that external callers will use command call parameterizations
+that can selectively ignore or act on result records, and that the process should therefore
+yield meaningful result records. If, in turn, the process itself receives a set of result
+records from a sub-process, these should be inspected individually in order to identify result
+values that could require re-annotation or status re-classification.
```

Comment:
Should there be a small paragraph on "message" and "hints" keys in the record are to be used to provide human-readable feedback to the user.  I think it would be valuable here as logging and ui talks about interfacing to the user and we do not even mention here how result records can be used for that.

Reply:
Added in fee7a8f7a107ffc44739935a84e4415acdf3f715